### ISSUE TITLE
Задание 5 №5

### ISSUE THREAD

**Issue by KimNikita** [SemyonBevzuk/Programming_practice_381908-1-2#48]
В условии задачи про кинотеатр говорится о VIP зоне, подразумевается VIP зал, или VIP место в зале?

**Comment by SemyonBevzuk** [comment]
VIP зал.

**Comment by KimNikita** [comment]
Какой тогда смысл пользователю указывать и номер зала и тип зоны, если тип зоны привязан к каждому конкретному залу? Или не привязан?

**Comment by SemyonBevzuk** [comment]
Так, сейчас посмотрю на какой-нибудь кинотеатр :D

**Comment by SemyonBevzuk** [comment]
Мда, ладно, давайте сделаем по заданию. Пусть у нас есть зал, в неё будет разделение на зоны. Какие именно места в зале VIP - решите сами.

Просто обычно же в кинотеатре отдельные VIP залы... Ну да ладно.

**Comment by KimNikita** [comment]
"принять заказ билетов покупателя: дату, время сеанса, название фильма, номер зала, тип зоны, число мест" - пользователь все сам вводит или выбирает из предложенного?

**Comment by KimNikita** [comment]
мне кажется что все кроме числа мест нужно выбирать из списка

**Comment by SemyonBevzuk** [comment]
Будем считать, что у пользователя есть View-модель, на которой производится визуализация доступных предложений. Он все вводит сам, вам нужно проверить корректность, если что-то не так - сообщить пользователю. Если данные корректны - обработать заказ.

**Comment by KimNikita** [comment]
ок

**Comment by SemyonBevzuk** [comment]
Если бы вы писали форму для кассы, то да. На экране был бы списки, которые брали бы данные из ядра. Но в самом ядре у вас бы была функция, которую вы должны реализовать. Пусть она получает данные с формы, которая уже как-то ограничивает ввод пользователя, и обрабатывает их.

**Comment by KimNikita** [comment]
![Безымянный](https://user-images.githubusercontent.com/60927460/79637326-2fc18080-8187-11ea-95f8-04dff9466702.jpg)
Это нормально? Я не писал перегрузки оператора присваивания.

**Comment by KimNikita** [comment]
Второй вопрос: пользователь может делать второй заказ, не закончив с первым? Тут больше вопрос к работе операции резервирования.

**Comment by KimNikita** [comment]
Операции 2-4 должны быть самостоятельными? По идее они должны вызываться в строгом порядке друг за другом после вызова операции 1

**Comment by SemyonBevzuk** [comment]
> ![Безымянный](https://user-images.githubusercontent.com/60927460/79637326-2fc18080-8187-11ea-95f8-04dff9466702.jpg)
> Это нормально? Я не писал перегрузки оператора присваивания.

Лучше напишите, иначе сработает оператор = по умолчанию, а он просто копирует значения полей. Если в классе есть работа с динамической памятью, то это приведет к созданию двух объектов, которые указывают на одну память. Эта ситуация может привести к проблемам при вызове деструкторов: первый объект удалиться корректно (но удалит память, на которую ссылается второй), а второй попробует удалить удаленную память и упадёт



> Второй вопрос: пользователь может делать второй заказ, не закончив с первым? Тут больше вопрос к работе операции резервирования.

1. В простом варианте - нет.
2. Более сложный вариант, но не идеальный, без учета id покупателя и его связи с заказом.
Есть три состояния мест в зале: свободно, продано, зарезервировано. Продать зарезервированные места мы можем только тогда, если на них пришел заказ (заказ полностью из зарезервированных мест). Предполагаем, что это просто покупка брони её заказчиком. Если в заказе есть другие места (свободные или занятые), то покупает другой пользователь, для него выдаем сообщение, что данные места заняты. Зарезервированные места при проверке количества мест считаем как занятые.
3. Осуществлять привязку покупателя к местам. Потребуется идентификатор покупателя, на которого оформляется заказ. Заказ может быть оплачен сразу, то есть билеты проданы. Или заказ с бронированием, то есть места меняют своё состояние на зарезервировано, для других пользователей они заняты, оплатить или отменить их может только заказчик. 


> Операции 2-4 должны быть самостоятельными? По идее они должны вызываться в строгом порядке друг за другом после вызова операции 1

Да, это отдельные самостоятельные методы, но они могут вызваться и в операции 1.
Пусть на основе вашей программы будет реализован сайт. Вполне допустимо, что пользователи захотят проверить количество свободных мест до покупки. Или захотят вычислить стоимость своей корзины с билетами. И операция резервирования тоже не обязывает пользователя сразу оплачивать билет.

**Comment by KimNikita** [comment]
Тогда на вход должны подаваться такие же данные как и у первой операции?

**Comment by SemyonBevzuk** [comment]
Да, можете даже под это создать отдельный класс или структуру типа Order.

**Comment by KimNikita** [comment]
![Безымянный](https://user-images.githubusercontent.com/60927460/80245317-22b6fc80-8673-11ea-85b5-576fce4c38dc.jpg)
У меня в каждой функции используется разное число параметров, если я создам структуру с полями от id до zone и буду использовать в каждой функции только нужные поля, это будет не очень, или нормально?

**Comment by SemyonBevzuk** [comment]
Если это сделать для всех функций, то это будет не очень хорошо.
Для FindSeance и CalculateCost это не нужно.

А вот TicketsOrder, IsFree, Reserve, CancelOrder, CreateTickets можно. Тут напрашивается некоторая сущность, которая может агрегировать данные о заказе.

**Comment by KimNikita** [comment]
Нужно делать такую же менюшку для работы с пользователем через консоль, как в FilmLibrary?

**Comment by SemyonBevzuk** [comment]
Вы можете или сделать консольное меню, или привести пример использования всех public методов.

**Comment by KimNikita** [comment]
При вводе из файла вылетает ошибка, если не обращаться к функции Date:: operator>> а ввести вручную, то работает, причем на Time:: operator>> не ругается, хотя там код аналогичный
![Безымянный2](https://user-images.githubusercontent.com/60927460/80490164-a02b8700-8969-11ea-9343-041fa3eca9b7.jpg)
![Безымянный3](https://user-images.githubusercontent.com/60927460/80490167-a1f54a80-8969-11ea-9bc7-a2da53a18c95.jpg)
![Безымянный](https://user-images.githubusercontent.com/60927460/80490169-a28de100-8969-11ea-936b-d197ec6fdee3.jpg)

**Comment by SemyonBevzuk** [comment]
А можно весь код посмотреть? Я не нашел его в вашем репозитории.

**Comment by KimNikita** [comment]
сделал реквест

**Comment by SemyonBevzuk** [comment]
Тут хитрая штука... Смотрите, у вас создаваемый объект Date создается по конструктору Date из прошлого задания. Видите, по памяти мы залезли вправо и записали параметры по умолчанию из прошлого задания, это видно по году первого фильма. 
![Безымянный](https://user-images.githubusercontent.com/10177297/80504641-0a4d2780-897c-11ea-91ee-6a3dc7a69dd1.png)
Но после спокойно работают методы текущего класса Date. Пока не могу сказать как это исправить... Видимо ошибка где-то при создании проекта, не тот файл подтянула студия.

**Comment by SemyonBevzuk** [comment]
Как я понял, у вас FilmLibrary из прошлого задания, она использует Film.h, в которой используется  Date.h из задания 4 (с годом фильма). Видимо при создании в задании 5 объекта Date используется первый попавшийся конструктор по умолчанию с подходящей семантикой, а это из Task4/Date.h, но после это ломает память, ибо текущий Date имеет только два поля.

**Comment by KimNikita** [comment]
Значит мне нужно удалить Date.h из предыдущего задания?

**Comment by SemyonBevzuk** [comment]
Их нужно как-то разделить, посмотрите внимательно как это сделать. Просто удалить не поможет, Task4/Date где-то же используется. Можно разделить по namespace-ам или просто дать этим классам разные имена.

**Comment by KimNikita** [comment]
Работает, я просто забыл filmlibrary из проекта убрать
